File: app/src/main/java/com/designpatterns/demo/singleton/java/ContainerSingleton.java
package com.designpatterns.demo.singleton.java;

import java.util.HashMap;
import java.util.Map;

/**
 * 单例模式 使用容器实现单例模式
 *
 * @author zhangzhiguo
 * @version 1.0.0
 * @email: 810903987@qq.com
 * @date 2018/4/27 上午9:52
 * @project_name DesignPatternsDemo
 */
public class ContainerSingleton {
    private static Map<String, Object> objMap = new HashMap<String, Object>();

    public static void registerService(String key, Object instance) {
        if (objMap.containsKey(key)) {
            //  添加单例模式
            objMap.put(key, instance);
        }

    }

    //   获取单例
    public static Object getService(String key) {
        return objMap.get(key);
    }
}

File: app/src/main/java/com/designpatterns/demo/creational/singleton/kotlin/ContainerSingleton.kt
package com.designpatterns.demo.creational.singleton.kotlin

import java.util.HashMap

/**
 * 单例模式 使用容器实现单例模式
 *
 * @author  zhangzhiguo
 * @email: 810903987@qq.com
 * @date 2018/4/27 上午9:58
 * @project_name DesignPatternsDemo
 * @version 1.0.0
 *
 */
class ContainerSingleton private constructor() {


    //Kotlin写法
    companion object {
        private lateinit var objMap: HashMap<String, Any>

        fun registerService1(key: String, instance: Any) {
            if (objMap.containsKey(key)) {
                objMap[key] = instance
            }
        }

        fun getService(key: String): Any {
            return objMap[key]!!
        }


    }


    //java转Kotlin
    private val objMap = HashMap<String, Any>()

    fun registerService(key: String, instance: Any) {
        if (objMap.containsKey(key)) {
            //  添加单例模式
            objMap[key] = instance
        }

    }

    //   获取单例
    fun getService(key: String): Any {
        return objMap[key]!!
    }
}